src/nvim/log.c

```diff
@@ -7,52 +7,74 @@
-#define USR_LOG_FILE_2 "$HOME" _PATHSEPSTR ".nvimlog"
-
-/// Cached location of the log file set by log_path_init()
-static char expanded_log_file_path[MAXPATHL + 1] = { 0 };
+/// Cached location of the expanded log file path decided by log_path_init().
+static char log_file_path[MAXPATHL + 1] = { 0 };
 
 static uv_mutex_t mutex;
 
 #ifdef INCLUDE_GENERATED_DECLARATIONS
 # include "log.c.generated.h"
 #endif
 
-/// Initialize path to log file
+static bool log_try_create(char *fname)
+{
+  if (fname == NULL || fname[0] == '\0') {
+    return false;
+  }
+  FILE *log_file = fopen(fname, "a");
+  if (log_file == NULL) {
+    return false;
+  }
+  fclose(log_file);
+  return true;
+}
+
+/// Initializes path to log file. Sets $NVIM_LOG_FILE if empty.
 ///
-/// Tries to use #USR_LOG_FILE, then falls back #USR_LOG_FILE_2. Path to log
+/// Tries $NVIM_LOG_FILE, or falls back to $XDG_DATA_HOME/nvim/log. Path to log
 /// file is cached, so only the first call has effect, unless first call was not
-/// successful. To make initialization not succeed either a bug in expand_env()
-/// is needed or both `$NVIM_LOG_FILE` and `$HOME` environment variables
-/// undefined.
+/// successful. Failed initialization indicates either a bug in expand_env()
+/// or both $NVIM_LOG_FILE and $HOME environment variables are undefined.
 ///
 /// @return true if path was initialized, false otherwise.
 static bool log_path_init(void)
 {
-  if (expanded_log_file_path[0]) {
+  if (log_file_path[0]) {
+    fprintf(stderr, "%ld WTFF=%s\n", os_get_pid(), log_file_path);
     return true;
   }
-  expand_env((char_u *)USR_LOG_FILE, (char_u *)expanded_log_file_path,
-             sizeof(expanded_log_file_path) - 1);
-  // if the log file path expansion failed then fall back to stderr
-  if (strcmp(USR_LOG_FILE, expanded_log_file_path) == 0) {
-    memset(expanded_log_file_path, 0, sizeof(expanded_log_file_path));
-    expand_env((char_u *)USR_LOG_FILE_2, (char_u *)expanded_log_file_path,
-               sizeof(expanded_log_file_path) - 1);
-    if (strcmp(USR_LOG_FILE_2, expanded_log_file_path) == 0) {
-      memset(expanded_log_file_path, 0, sizeof(expanded_log_file_path));
+  size_t size = sizeof(log_file_path);
+  expand_env((char_u *)"$" LOG_FILE_ENV, (char_u *)log_file_path,
+             (int)size - 1);
+  fprintf(stderr, "%ld WTF=%s\n", os_get_pid(), log_file_path);
+  if (strequal("$" LOG_FILE_ENV, log_file_path)
+      || log_file_path[0] == '\0'
+      || os_isdir((char_u *)log_file_path)
+      || !log_try_create(log_file_path)) {
+    // Invalid $NVIM_LOG_FILE or failed to expand; fall back to default.
+    memset(log_file_path, 0, size);
+    char *defaultpath = stdpaths_user_data_subpath("log", 0, true);
+    size_t len = xstrlcpy(log_file_path, defaultpath, size);
+    xfree(defaultpath);
+    // Fall back to .nvimlog
+    if (len >= size || !log_try_create(log_file_path)) {
+      len = xstrlcpy(log_file_path, ".nvimlog", sizeof(".nvimlog"));
```

Comment:
`sizeof(".nvimlog")` is not the size that is expected here, it should be the size of the destination buffer.